File: Code/Scala/NumberClassifier.scala
object NumberClassifierFunctional {

  def isFactor(number: Int, factor: Int) =
    number % factor == 0

  def factors(number : Int) =
    (1 to number) filter (number % _ == 0)

  def sum(factors: Seq[Int]) =
    factors.foldLeft(0)(_ + _)

  def isPerfect(number: Int) =
    sum(factors(number)) - number == number

  def isAbundant(number: Int) =
    sum(factors(number)) - number > number

  def isDeficient(numner: Int) =
    sum(factors(number)) - number < number

}

File: WhyFpJvm/src/com/company/NumberClassifier.java
package com.company;

import java.lang.Integer;import java.util.Collection;
import java.util.Collections;
import java.util.HashSet;
import java.util.Set;

// all methods accept number as a parameter, internal state exists
// all methods are public static because they are pure functions
// inefficient for repeating classifications (no caching)

public class NumberClassifier {

    public static boolean isFactor(final int candidate, final int number) {
        return number % candidate == 0;
    }

    public static Set<Integer> factors(final int number) {
        Set<Integer> factors = new HashSet<Integer>();
        factors.add(1);
        factors.add(number);
        for (int i = 2; i < number; i++)
            if (isFactor(i, number))
                factors.add(i);
        return factors;
    }




    public static int aliquotSum(final Collection<Integer> factors) {
        int sum = 0;
        int targetNumber = Collections.max(factors);
        for (int n : factors) {
            sum += n;
        }
        return sum - targetNumber;
    }

    public static boolean isPerfect(final int number) {
        return aliquotSum(factors(number)) == number;
    }

    public static boolean isAbundant(final int number) {
        return aliquotSum(factors(number)) > number;
    }

    public static boolean isDeficient(final int number) {
        return aliquotSum(factors(number)) < number;
    }
}
